koku/api/provider/serializers.py

```diff
@@ -60,7 +63,61 @@ class Meta:
         """Metadata for the serializer."""
 
         model = ProviderBillingSource
-        fields = ('__all__')
+        fields = ('uuid', 'bucket')
+
+
+def _get_sts_access(provider_resource_name):
+    """Get for sts access."""
+    # create an STS client
+    sts_client = boto3.client('sts')
+
+    # Call the assume_role method of the STSConnection object and pass the role
+    # ARN and a role session name.
+    assumed_role = sts_client.assume_role(
+        RoleArn=provider_resource_name,
+        RoleSessionName='AccountCreationSession'
+    )
+    credentials = assumed_role.get('Credentials')
+    access_key_id = credentials.get('AccessKeyId') if credentials else None
+    secret_access_key = credentials.get('SecretAccessKey') if credentials else None
+    session_token = credentials.get('SessionToken') if credentials else None
+
+    return (access_key_id, secret_access_key, session_token)
+
+
+def _check_s3_access(access_key_id, secret_access_key,
+                     session_token, bucket):
+    """Check for access to s3 bucket."""
+    s3_exists = True
+    s3_resource = boto3.resource(
+        's3',
+        aws_access_key_id=access_key_id,
+        aws_secret_access_key=secret_access_key,
+        aws_session_token=session_token,
+    )
+    try:
+        s3_resource.meta.client.head_bucket(Bucket=bucket)
+    except (botocore.exceptions.ClientError,
+            requests.exceptions.ConnectionError):
+        s3_exists = False
+    return s3_exists
+
+
+def _check_org_access(access_key_id, secret_access_key, session_token):
+    """Check for provider organization access."""
+    access_ok = True
+    org_client = boto3.client(
+        'organizations',
+        aws_access_key_id=access_key_id,
+        aws_secret_access_key=secret_access_key,
+        aws_session_token=session_token,
+    )
+    try:
+        org_client.describe_organization()
```

Comment:
This is future-thinking out loud, but once we've decided what our role policy is we might want to expand out this section of checks to make sure each permission in the policy is checked (e.g. instead of just checking for the bucket, see if we can add an event listener).

I still don't have a great view in my head of what we want to automate vs. what we're asking the customer to do as manual on-boarding in their AWS account. As an example, I know we need them to add an IAM role so that we can use assume_role, but will we ask them to setup the cost usage report, or are we trying to do that? I know this is outlined in #102 but maybe we can discuss this and produce the role JSON so we know explicitly what access will ask for.

Reply:
Understood. I have less AWS experience and am trying to learn this stuff on the fly. I created a separate issue for the API, because we need to document what we are asking the user to do (but maybe we can also make it better over time).